vsintegration/src/FSharp.Editor/CodeFixes/ChangePrefixNegationToInfixSubtraction.fs

```diff
@@ -2,40 +2,45 @@
 
 namespace Microsoft.VisualStudio.FSharp.Editor
 
-open System
 open System.Composition
-open System.Threading.Tasks
 open System.Collections.Immutable
+open System.Text.RegularExpressions
 
 open Microsoft.CodeAnalysis.Text
 open Microsoft.CodeAnalysis.CodeFixes
 
+open CancellableTasks
+
 [<ExportCodeFixProvider(FSharpConstants.FSharpLanguageName, Name = CodeFix.ChangePrefixNegationToInfixSubtraction); Shared>]
-type internal ChangePrefixNegationToInfixSubtractionodeFixProvider() =
+type internal ChangePrefixNegationToInfixSubtractionCodeFixProvider() =
     inherit CodeFixProvider()
 
     static let title = SR.ChangePrefixNegationToInfixSubtraction()
 
-    override _.FixableDiagnosticIds = ImmutableArray.Create("FS0003")
-
-    override _.RegisterCodeFixesAsync context : Task =
-        asyncMaybe {
-            let! sourceText = context.Document.GetTextAsync(context.CancellationToken)
-
-            let mutable pos = context.Span.End + 1
-
-            // This won't ever actually happen, but eh why not
-            do! Option.guard (pos < sourceText.Length)
-
-            let mutable ch = sourceText.[pos]
-
-            while pos < sourceText.Length && Char.IsWhiteSpace(ch) do
-                pos <- pos + 1
-                ch <- sourceText.[pos]
-
-            // Bail if this isn't a negation
-            do! Option.guard (ch = '-')
-            do context.RegisterFsharpFix(CodeFix.ChangePrefixNegationToInfixSubtraction, title, [| TextChange(TextSpan(pos, 1), "- ") |])
-        }
-        |> Async.Ignore
-        |> RoslynHelpers.StartAsyncUnitAsTask(context.CancellationToken)
+    override _.FixableDiagnosticIds = ImmutableArray.Create "FS0003"
+
+    override this.RegisterCodeFixesAsync context = context.RegisterFsharpFix this
+
+    interface IFSharpCodeFixProvider with
+        member _.GetCodeFixIfAppliesAsync context =
+            cancellableTask {
+                let! sourceText = context.GetSourceTextAsync()
+
+                // in a line like "... x  -1 ...",
+                // squiggly goes for "x", not for "-", hence we search for "-"
+                let remainingText = $"{sourceText.GetSubText(context.Span.End)}"
```

Comment:
$"{}" should never be a replacement for .ToString() in such case.
It's both worse performance wise as well as worse for readability (reader has to stop and figure out why the hell is string interpolation needed here).

Reply:
Correction, even worse I guess - the previous code was carefully reading one character at a time from the Roslyn's SourceText.
The new code allocates for the full substring from position till end of the document, even though the algorithm itself only cares about immediate whitespaces after the position.

Reply:
Readability is debatable, performance - I guess not. I will fix this in the followup soon!

On the side note - if this is a universal rule, would it make sense to start producing some diag here?